File: Lesson12_Inheritance/05_AbstractClasses/MyExample.py
"""
Copyright Khawla Shnaikat and ChatGPT 3.5, 2023
Licensed under GPL v3
See LICENSE.txt for more information.
"""
from abc import ABC, abstractmethod

class Animal(ABC):
    def eat(self):
        print("Method eat is supplied by Animal")

class Real(Animal, ABC):
    def is_real(self):
        print("Method isReal is supplied by Real")

class Swimmer(Real, ABC):
    def swim(self):
        print("Method swim is supplied by Swimmer")

    @abstractmethod
    def breathe(self):
        pass

class Trilobite(Swimmer):
    def is_extinct(self):
        print("Method isExtinct is supplied by Trilobite")

    def breathe(self):
        print("Method breathe is supplied by Trilobite")

def main():
    my_trilobite = Trilobite()
    my_trilobite.swim()
    my_trilobite.eat()
    my_trilobite.is_extinct()
    my_trilobite.is_real()
    my_trilobite.breathe()

    # The following code would raise a TypeError as Swimmer is an abstract class
    # my_swimmer = Swimmer()
    # my_swimmer.swim()

if __name__ == "__main__":
    main()

File: Lesson12_Inheritance/05_AbstractClasses/MyExample.java
/*
Copyright Ann Barcomb and Emily Marasco, 2023
Licensed under GPL v3
See LICENSE.txt for more information.
*/
package edu.ucalgary.oop;

abstract class Animal {
	public void eat() {
		System.out.println("Method eat is supplied by Animal");
	}
}

abstract class Real extends Animal {
	public void isReal() {	
		System.out.println("Method isReal is supplied by Real");
	}
}

abstract class Swimmer extends Real {
	public void swim() {
		System.out.println("Method swim is supplied by Swimmer");
	}

	public abstract void breathe();
}


final class Trilobite extends Swimmer {
	public void isExtinct() {
		System.out.println("Method isExtinct is supplied by Trilobite");
	}

        @Override
	final public void breathe() {
		System.out.println("Method breathe is supplied by Trilobite");
	}
}

public class MyExample {
	public static void  main(String[] args) {
		Trilobite myTrilobite = new Trilobite();
		myTrilobite.swim();
		myTrilobite.eat();
		myTrilobite.isExtinct();
		myTrilobite.isReal();
		myTrilobite.breathe();


		/* The following code won't work as the class is abstract.
		var mySwimmer = new Swimmer();
		mySwimmer.swim();
		mySwimmer.eat();
		*/
	}


}
